### DISCUSSION TITLE
Running multiple external controllers for the same robot

### DISCUSSION THREAD

**Discussion by tonynajjar** [cyberbotics/wrestling#20] [Q&A] [answered]
Hello, I'm new to Webots and participating in the wrestling competition. I'm using ROS so I'm creating a distributed system of Nodes. In many of those nodes I need to access _controllers.Robot_ (e.g to retrieve sensor data or to actuate motors). However when I try to do this I get:

`INFO: 'participant' extern controller: refusing connection attempt from another extern controller.`

Is there a way to run multiple external controllers for the same robot?

I found this link but it doesn't answer my question:

https://cyberbotics.com/doc/guide/running-extern-robot-controllers#single-simulation-and-multiple-local-extern-robot-controllers - This talks about multiple controllers for multiple robots

**Comment by omichel** [comment]
Actually, if you want to use ROS, you should not use the Webots controller API at all.
Instead, you should use the [webots_ros2](https://github.com/cyberbotics/webots_ros2) interface.
In webots_ros2, a launcher usually start both Webots and your ROS nodes.
You should therefore simplify this launcher: remove the launch of Webots and keep only the launch of your ROS controller node (and all the other ROS nodes that you need). Before starting this launcher, you should set the [WEBOTS_CONTROLLER_URL](https://github.com/cyberbotics/wrestling/blob/a8e1054063fbaa3d40eb2cf357b8c312041d8c97/controllers/Dockerfile#L18-L19) environment variable (as it is currently set in the default Dockerfile) which is defined by the CI, so that your ROS controller will be able to connect with Webots.
Here is an example of launcher: https://github.com/cyberbotics/webots_ros2/blob/master/webots_ros2_epuck/launch/robot_launch.py
This one is for an e-puck robot and launches Webots as well. So, you can modify it to adapt it to your need and in particular remove the launch of Webots from it. This is about [here](https://github.com/cyberbotics/webots_ros2/blob/a77c4ae5a84bd279db395103d0c008329cf3f1c5/webots_ros2_epuck/launch/robot_launch.py#L121-L143) in the launcher. You should be able to test all this locally on your Linux machine before pushing it to the CI.
Let me know if you have any further question while doing this.

**Reply by rua0ra1** [reply]
Can I use same webots_ros2 interface for ROS1 too?

**Reply by omichel** [reply]
No. For ROS 1, you should use the [webots_ros](http://wiki.ros.org/webots_ros) interface.

**Reply by rua0ra1** [reply]
Thank you for the info @omichel

**Reply by tonynajjar** [reply]
Hi @omichel thanks for the answer and sorry for my late answer. The problem with the webots_ros2 interface is that it doesn't give me everything I need. For example, I don't see the gyro/accelerometer values published although they are part of the model. So to compensate for that I was hoping to write my own nodes that access controllers.Robot to get the gyro values from there and publish them to the other nodes. However when I try to do this I get the error I mentioned before.

I guess I'll have to implement a new plugin in webots_ros2_driver? Any other way?

**Reply by tonynajjar** [reply]
> I guess I'll have to implement a new plugin in webots_ros2_driver?

Had a quick look, it's not super straightforward, e.g. why are the plugins using the C interface of the sensors instead of the CPP interface? (e.g `Ros2LightSensor.hpp` imports` c/webots/light_sensor.h` instead of `cpp/webots/LightSensor.hpp`)

**Reply by tonynajjar** [reply]
Okay I missed the tutorial obviously: https://docs.ros.org/en/galactic/Tutorials/Advanced/Simulators/Webots.html.

It seems like what I'm looking for but I'm running through another issue. The driver loads for a second but then crashes

in webots:

```
INFO: 'participant' extern controller: connected.
INFO: 'participant' extern controller: refusing connection attempt from another extern controller.
INFO: 'participant' extern controller: disconnected, waiting for new connection.
```

in terminal:

`[ERROR] [driver-2]: process has died [pid 65394, exit code 1, cmd '/root/ros2/install/webots_ros2_driver/lib/webots_ros2_driver/driver --ros-args --params-file /tmp/launch_params_ynla4djg'].
`

I'm almost sure that the problem is not in my code because it reproduces with the the package of the tutorial I linked above. 

I'm running `vulcanexus:humble-desktop` and have webots_ros2 built from source, I tried both master and 2023.0.1.

**Reply by tonynajjar** [reply]
> I'm almost sure that the problem is not in my code because it reproduces with the the package of the tutorial I linked above.

more confirmation, it reproduces when I run the tests of webots_ros2_tests

```
INFO: 'Pioneer_3_AT' extern controller: connected.
INFO: 'Pioneer_3_AT' extern controller: refusing connection attempt from another extern controller.
INFO: 'Pioneer_3_AT' extern controller: disconnected, waiting for new connection.
```

Let me know if I should file an issue in the repo

**Reply by Jean-Eudes-le-retour** [reply]
I've done some tests, it appears that the Docker images from eprosima/vulcanexus use an old version of webots_ros2 (2023.0.0) which have a problem with the LD_LIBRARY_PATH environment variable.

You can patch it by adding `ENV LD_LIBRARY_PATH=/opt/ros/humble/lib/controller:$LD_LIBRARY_PATH` in your Dockerfile and it should fix some problems. Then normally you shouldn't need to reinstall webots_ros2 on top of the one present in the Docker image. Let us know if that helped!

**Reply by ygoumaz** [reply]
I am currently able to locally run all the tests on the master branch successfully, indicating that the issue is most likely related to your configuration. Which version of Webots are you using? 

The error message you get shows that 2 different controllers are trying to connect to the same robot (in the test the 'Pioneer_3_AT'). Are you using the following command to start the test from the workspace?
```
launch_test src/webots_ros2/webots_ros2_tests/test/test_system_driver.py
```

If so, could you provide the full logs you get in this case (in addition to the 3 lines you gave in your previous comment)?

**Reply by tonynajjar** [reply]
Thanks for your answers, here is more info about my setup:

```
root@logi-XMG-CORE-REN-M20:~/ros2/src# launch_test webots_ros2/webots_ros2_tests/test/test_system_driver.py
[INFO] [launch]: All log files can be found below /root/.ros/log/2023-02-03-11-04-05-342607-logi-XMG-CORE-REN-M20-32349
[INFO] [launch]: Default logging verbosity is set to INFO
testCamera (test_system_driver.TestDriver) ... [INFO] [webots-1]: process started with pid [32351]
[INFO] [ros2_supervisor.py-2]: process started with pid [32357]
[INFO] [driver-3]: process started with pid [32359]
[driver-3] Cannot connect to Webots instance, retrying in 0 second...
[ros2_supervisor.py-2] Cannot connect to Webots instance on socket "/tmp/webots_npxyacmyxx/webots-1234/ipc/Ros2Supervisor/extern", retrying in 0 second...
[webots-1] WARNING: It is not recommended to run Webots as root.
[driver-3] Cannot connect to Webots instance on socket "/tmp/webots_npxyacmyxx/webots-1234/ipc/Pioneer_3_AT/extern", retrying in 1 second...
[ros2_supervisor.py-2] Cannot connect to Webots instance on socket "/tmp/webots_npxyacmyxx/webots-1234/ipc/Ros2Supervisor/extern", retrying in 1 second...
[webots-1] WARNING: Cannot initialize the sound engine: Cannot initialize OpenAL default device 'OpenAL Soft'
[webots-1] INFO: 'Pioneer_3_AT' extern controller: waiting for connection on ipc://1234/Pioneer_3_AT or on tcp://<ip_address>:1234/Pioneer_3_AT
[webots-1] INFO: 'Ros2Supervisor' extern controller: waiting for connection on ipc://1234/Ros2Supervisor or on tcp://<ip_address>:1234/Ros2Supervisor
[webots-1] INFO: 'Pioneer_3_AT' extern controller: connected.
[webots-1] INFO: 'Ros2Supervisor' extern controller: connected.
[webots-1] INFO: 'Pioneer_3_AT' extern controller: refusing connection attempt from another extern controller.
[webots-1] INFO: 'Pioneer_3_AT' extern controller: disconnected, waiting for new connection.
[ERROR] [driver-3]: process has died [pid 32359, exit code 1, cmd '/root/ros2/install/webots_ros2_driver/lib/webots_ros2_driver/driver --ros-args --params-file /tmp/launch_params_ihgfystj'].
```

```
root@logi-XMG-CORE-REN-M20:~/ros2/src# webots --version
Webots version: R2023a
```

```
root@logi-XMG-CORE-REN-M20:~/ros2/src/webots_ros2# git status
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
```

```
root@logi-XMG-CORE-REN-M20:~/ros2/src/webots_ros2# echo $LD_LIBRARY_PATH
/opt/ros/humble/lib/controller:/root/ros2/install/webots_ros2_control/lib:/root/ros2/install/webots_ros2_driver/lib:/root/ros2/install/webots_ros2_msgs/lib:/opt/ros/humble/opt/rviz_ogre_vendor/lib:/opt/ros/humble/lib/x86_64-linux-gnu:/opt/ros/humble/lib:/usr/local/webots/lib/controller/:/lib/controller:
```

> it appears that the Docker images from eprosima/vulcanexus use an old version of webots_ros2 (2023.0.0)

I do build it from source on master though.

@omichel has access to my dockerfile, if you would like to try to reproduce I guess he can share it with you. I'll have more time to debug today night or tomorrow

**Reply by ygoumaz** [reply]
I may have found the origin of the problem. I suspect that the libController used by the driver library is not the correct one. 

In your Docker container, there is two versions of `webots_ros2`, the one installed by default in `/opt/ros/humble` (which is `2023.0.0`) and the one you installed from sources in `~/ros2/src` (which points to master). The two installations seem to be in conflict, as some logs you get can only come from the other installation (`2023.0.0`), as they have been changed in master and `2023.0.1`. 

You can remove `/opt/ros/humble/lib/controller` from your `LD_LIBRARY_PATH` and also uninstall the `webots_ros2` version contained in `/opt/ros/humble` by running `sudo apt remove ros-humble-webots-ros2`.

This conflict can cause the driver library to crash, resulting in the controller respawning and the new (rejected) connection attempt.

Please let me know if this solves your problem. If not, could you provide the same logs after applying the above solution?

**Reply by tonynajjar** [reply]
Here is what I did: 

- change from the vulcanexus image to `osrf/ros:humble-desktop`
- install webots and webots_ros2 via apt
- I have these environment variables in the dockerfile:

```
ENV WEBOTS_HOME /usr/local/webots
ENV PATH /usr/local/webots:${PATH}
ENV PYTHONPATH=${WEBOTS_HOME}/lib/controller/python
ENV PYTHONIOENCODING=UTF-8
ENV LD_LIBRARY_PATH=${WEBOTS_HOME}/lib/controller:${LD_LIBRARY_PATH}
ARG WEBOTS_CONTROLLER_URL
ENV WEBOTS_CONTROLLER_URL=${WEBOTS_CONTROLLER_URL}
```

and I still get the same error... it's mind-boggling

```
root@logi-XMG-CORE-REN-M20:~/ros2/src# ros2 launch webots_wrestler simulation.py
[INFO] [launch]: All log files can be found below /root/.ros/log/2023-02-03-21-09-19-674367-logi-XMG-CORE-REN-M20-82065
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [webots-1]: process started with pid [82066]
[INFO] [driver-2]: process started with pid [82068]
[INFO] [ros2_supervisor.py-3]: process started with pid [82070]
[driver-2] Cannot connect to Webots instance, retrying in 0 second...
[ros2_supervisor.py-3] Cannot connect to Webots instance on socket "/tmp/webots-1234/ipc/Ros2Supervisor/extern", retrying for another 50 seconds...
[driver-2] Cannot connect to Webots instance on socket "/tmp/webots-1234/ipc/participant/extern", retrying in 1 second...
[driver-2] Cannot connect to Webots instance on socket "/tmp/webots-1234/ipc/participant/extern", retrying in 2 seconds...
[ERROR] [driver-2]: process has died [pid 82068, exit code 1, cmd '/root/ros2/install/webots_ros2_driver/lib/webots_ros2_driver/driver --ros-args --params-file /tmp/launch_params_ot45fq2i'].
```

This error is from own driver but also reproduces with the test from webots_ros2_tests

Here is a few information about the new container:

```
root@logi-XMG-CORE-REN-M20:~/ros2/src# echo $LD_LIBRARY_PATH 
/opt/ros/humble/opt/rviz_ogre_vendor/lib:/opt/ros/humble/lib/x86_64-linux-gnu:/opt/ros/humble/lib:/usr/local/webots/lib/controller:

root@logi-XMG-CORE-REN-M20:~/ros2/src# webots --version
Webots version: R2023a
```
webots_ros2 is not installed from source anymore.

**Reply by tonynajjar** [reply]
Maybe this helps you:

```
root@logi-XMG-CORE-REN-M20:~/ros2/src# ros2 run --prefix 'gdb -ex run --args' webots_ros2_driver driver --ros-args --params-file /tmp/launch_params_kp8h8c7_
GNU gdb (Ubuntu 12.1-0ubuntu1~22.04) 12.1
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /opt/ros/humble/lib/webots_ros2_driver/driver...
(No debugging symbols found in /opt/ros/humble/lib/webots_ros2_driver/driver)
Starting program: /opt/ros/humble/lib/webots_ros2_driver/driver --ros-args --params-file /tmp/launch_params_kp8h8c7_
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff60d4640 (LWP 17083)]
[New Thread 0x7ffff58d3640 (LWP 17084)]
[New Thread 0x7ffff504b640 (LWP 17085)]
[New Thread 0x7ffff484a640 (LWP 17086)]
[New Thread 0x7ffff4049640 (LWP 17087)]
[New Thread 0x7ffff3848640 (LWP 17088)]
[New Thread 0x7ffff3047640 (LWP 17089)]
[New Thread 0x7ffff2839640 (LWP 17090)]
[New Thread 0x7ffff1f58640 (LWP 17091)]
[New Thread 0x7ffff16bb640 (LWP 17092)]
[New Thread 0x7ffff0e45640 (LWP 17093)]
[Thread 0x7ffff484a640 (LWP 17086) exited]
[Thread 0x7ffff4049640 (LWP 17087) exited]
[Thread 0x7ffff3848640 (LWP 17088) exited]
[Thread 0x7ffff2839640 (LWP 17090) exited]
[Thread 0x7ffff3047640 (LWP 17089) exited]
[Thread 0x7ffff504b640 (LWP 17085) exited]
[Thread 0x7ffff1f58640 (LWP 17091) exited]
[Thread 0x7ffff58d3640 (LWP 17084) exited]
[Thread 0x7ffff60d4640 (LWP 17083) exited]
[Thread 0x7ffff0e45640 (LWP 17093) exited]
[Thread 0x7ffff6e96f40 (LWP 17074) exited]
[Thread 0x7ffff16bb640 (LWP 17092) exited]
[New process 17074]
[Inferior 1 (process 17074) exited with code 01]
```

**Reply by tonynajjar** [reply]
I localized where it crashes :detective: , it's at this [line](https://github.com/cyberbotics/webots_ros2/blob/4a4a56bfdf12ad745118c7e1b40f8251b1a393c5/webots_ros2_driver/src/PythonPlugin.cpp#L67)

**Reply by ygoumaz** [reply]
From what I see, it is still the wrong libController that is used by your ROS package. Is the bug still occurring if you remove the following environment variable from your Dockerfile ? 

```
ENV LD_LIBRARY_PATH=${WEBOTS_HOME}/lib/controller:${LD_LIBRARY_PATH}
```

`LD_LIBRARY_PATH`  is used to get the path to the Webots libController only if you use C/C++ or (sometimes) Python extern controllers. Same goes with `PYTHONPATH` and `PYTHONIOENCODING`, which are only useful if you use Python extern controllers. These environment variables may infer with the ROS package and should be unset if you are using the ROS package.

**Reply by Jean-Eudes-le-retour** [reply]
I've made a simple working ROS 2 example, you can check out the code here: https://github.com/cyberbotics/wrestling-alice-ros-2

**Comment by omichel** [comment, marked as answer]
The [Vulcanexus](https://vulcanexus.org) ROS distribution provides a fairly lightweight [humble-simulation](https://hub.docker.com/r/eprosima/vulcanexus/tags) docker image that includes the `webots_ros2` package. That should be very convenient to use in the wrestling competition.

The [Alice (ROS)](https://github.com/cyberbotics/wrestling-alice-ros-2) example provide a complete template based on Vulcanexus for participants willing to use ROS for their robot controller.
